hyperspy/io_plugins/nexus.py

```diff
@@ -0,0 +1,1251 @@
+    as MAX_SIZE.  Chunks are generally close to some power-of-2 fraction of
+    each axis, slightly favoring bigger values for the last index.    
+    
+    Parameters
+    ----------
+    data  -  numpy, dask array or h5py.Dataset
+    
+    Returns
+    -------
+    tuple 
+        Estimated chunk sizes for each axis
+    
+    """
+    chunks= h5py._hl.filters.guess_chunk(data.shape, None, np.dtype(data.dtype).itemsize)
+    return chunks
+
+
+def _write_nexus_groups(dictionary, group, **kwds):
+    """Recursively iterate through dictionary and 
+    create the groups and datasets
+    
+    Parameters
+    ----------
+    dictionary : dict
+        dictionary contents to store to hdf group
+    group : hdf group
+        location to store dictionary
+    **kwds : additional keywords
+       additional keywords to pass to h5py.create_dataset method
+       
+    """
+    for key, value in dictionary.items():
+        if key == 'attrs':
+            # we will handle attrs later...
+            continue
+        if isinstance(value, dict):
+            if "attrs" in value:
+                if "NX_class" in value["attrs"] and value["attrs"]["NX_class"]=="NXdata":
+                    continue
+            if 'value' in value.keys() and not isinstance(value["value"],dict) \
+            and len(set(list(value.keys())+["attrs","value"]))==2:
+               value=value["value"]
+            else:
+                _write_nexus_groups(value, group.require_group(key),
+                      **kwds)
+        
+        if isinstance(value, (list, tuple,np.ndarray, da.Array)):
+            data = _parse_to_file(value)
+            overwrite_dataset(group, data, key, chunks=None, **kwds)
+        elif isinstance(value, (int,float,str,bytes)):
+            group.create_dataset(key,
+                data=_parse_to_file(value))     
+
+
+
+def _write_nexus_attr(dictionary, group):
+    """Recursively iterate through dictionary and 
+    write "attrs" dictionaries
+    This step is called after the groups and datasets
+    have been created
+    
+    Parameters
+    ----------
+    dictionary : dict  
+        Input dictionary to be written to the hdf group
+    group : hdf group 
+    
+    """
+    for key, value in dictionary.items():
+        if key == 'attrs':
+
+            for k, v in value.items():
+                group.attrs[k] = _parse_to_file(v)        
+        else:
+            if isinstance(value, dict):
+                if "attrs" in value:
+                    if "NX_class" in value["attrs"] and value["attrs"]["NX_class"]=="NXdata":
+                        continue
+                _write_nexus_attr(dictionary[key], group[key])
+                
+def _guess_signal_type(data):   
+    """Checks last 2 dimensions
+    An area detector will be square, > 255 pixels each side and
+    if rectangular ratio of sides is at most 4.
+    An EDS detector will generally by ndetectors*2048 or ndetectors*4096
+    the ratio test will be X10-1000  
```

Comment:
Ok have removed.   Nexus does not actually store any information on how data should be plotted or presented - it stores the data and axes. I've added  attributes so that when a file is saved the navigation details are stored with the axis data.

Reply:
Is there anything in the metadata, which can be used to determine if this should be a signal1d or signal2d?
Detector type, acquisition mode, etc.